### DISCUSSION TITLE
Download server binaries on start or include in container?

### DISCUSSION THREAD

**Discussion by LsHallo** [LsHallo/ets2-dedicated-convoy-server#1] [Polls]
Currently I'm downloading and installing the ETS/ATS server during build time. This means I have to rebuild every time they release a new version and you would have to wait for me to notice and trigger a build. Maybe this could be automated?

Would it be nicer to let the server dynamically download the server executable every time you start it and it can't find the executables already installed?
This would mean longer first start up time and may lead to other problems when they change the file structure in the server executable and move e.g. the binary file. This would break the docker file for everyone and you can't keep using the old version of the server.
It would also allow independent updates of the server by simply recreating it (provided the server data is not mounted on the host and persistent).

Pro Download On Start:
- Will download newer version when available.
- Independent updates without waiting for a new docker container.
- No possible legal trouble for distributing ETS code + binaries.
- Smaller container size (download).
- Updates possible without waiting for new image to be built.

Con Download on Start:
- Longer first startup.
- Confusion when the server takes 10+ minutes on slow connections.
- Version tags on the container image become a bit meaningless. As does pinning to a version when running.

Would love to get some opinions!

**Comment by joaogenio** [comment]
This may be slighlty out of topic but what happens if the server binaries (eventually) get outdated? If players update their game, will they be able to join the server? This could either be considered a Pro or a Con for the poll.

Also, the fact that doing this dynamic download can lead to other problems (as you mentioned) should be considered a Con. People may look at the Pros/Cons as a TLDR and vote without taking this into consideration.

**Reply by LsHallo** [reply]
That's what i was trying to hint at that you may have to wait for me to build a new image when a new version releases.

**Reply by joaogenio** [reply]
If SCS launches a new server binary and it messes with the file structure (e.g.), then you will always have to update the container **independently** of the poll choice. The first choice will break with every update from SCS. Both will break with an "abnormal" update from SCS.

**Reply by LsHallo** [reply]
Yes that is true. What might be advantageous for the "have server files in container" option is that you could run an older version without having to worry about accidentally updating your server. I have no clue if that's possible and the matchmaking servers will accept an older version to be listed, but it could be worthwhile to keep it like it is right now until we know for sure. It could be nice if there are mod incompatibilities and you want to stay on an older version.

**Comment by LsHallo** [comment]
As this discussion seems pretty conclusive I've opened #2. Take a look if you have a few minutes. I might have missed something ;)
It introduces `ETS_SERVER_UPDATE_ON_START` env variable for full control. When no server files are found the update is performed regardless of the variable setting.
